File: 1710. Maximum Units on a Truck.py
class Solution:
    def maximumUnits(self, boxTypes, truckSize):
        boxTypes.sort(key=lambda x: -x[1])
        ans = 0
        for t, u in boxTypes:
            if truckSize > t:
                truckSize -= t
                ans += t * u
            else:
                ans += truckSize * u
                return ans
        return ans
        """
        #sorting algorithm for the box type capacity
        #algorithm looks for largest box type and puts it to the front
        maximum = 0
        for biggest in range(0,len(boxTypes)-1):
            for thickest in range(0,len(boxTypes)-1):
                if boxTypes[thickest][1] < boxTypes[thickest+1][1]:
                    maximum = thickest+1
            boxTypes.insert(biggest,boxTypes.pop(maximum))
 
        units = 0
        #and then keep adding until you hit the box limit
        while truckSize>0:
            if boxTypes[0][0] == 0:
                boxTypes.pop(0)
                continue
            units += boxTypes[0][1]
            boxTypes[0][0] += -1
        return units
        """

File: 1710. Maximum Units on a Truck.java
class Solution {
    public int maximumUnits(int[][] B, int T) {
        Arrays.sort(B, (a,b) -> b[1] - a[1]);
        int ans = 0;
        for (int[] b : B) {
            int count = Math.min(b[0], T);
            ans += count * b[1];
            T -= count;
			if (T == 0) return ans;
        }
        return ans;
    }
}
/*
class Solution {
    public int maximumUnits(int[][] boxTypes, int truckSize) {

        //a sorting algorithm to start with
        Arrays.sort(boxTypes, new Comparator<int[]>() {
            public int compare(int[] boxType1, int[] boxType2) {
                if (boxType1[1] != boxType2[1])
                    return boxType2[1] - boxType1[1];
                else
                    return boxType2[0] - boxType1[0];
            }
        });
        
        //keep adding up until the capacity is lower than the chunk amount
        int answer = 0;
        for(int index=0;index<boxTypes.length;index++){
            if(boxTypes[index][0] > truckSize){
                answer+= boxTypes[index][0]*boxTypes[index][1];
                truckSize-= boxTypes[index][0];
            }
            else{
                answer+= boxTypes[index][1]*truckSize;
                break;
            }
        }
        return answer;
    }
}*/